chain/client/src/sync_jobs_actions.rs

```diff
@@ -81,6 +81,9 @@ impl SyncJobsActions {
     }
 
     pub fn handle_apply_state_parts_request(&mut self, msg: ApplyStatePartsRequest) {
+        // Unload mem-trie (in case it is still loaded) before we apply state parts.
+        msg.runtime_adapter.unload_mem_trie(&msg.shard_uid);
```

Comment:
How does this interact with the retain_mem_tries earlier? Do I understand this correctly:

- If in epoch T-1 we track 1, in epoch T we track 2, and in epoch T+1 we track 3, then in epoch T, we retain [2, 3] (unload 1), and here in state sync we unload 3 (in case it's still loaded)
- If in epoch T-1 we track 1, in epoch T we track 2, and in epoch T+1 we track 1, then in epoch T, we retain [2, 1] (unload nothing), and here in state sync we unload 1

If that understanding is correct, would it work instead to just call retain_mem_tries using only the "*this* epoch cares" shards, and then omit the unload_mem_trie call here? Or is there some issue with that?

Reply:
Your understanding is correct.
The reason I added `unload_mem_trie` is some unit test failed without it, because it ran state sync for a shard it already tracked. It looks like a bug or impossible situation in production, but I added `unload_mem_trie` just in case.